cpp/include/raft/spatial/knn/detail/ivf_pq_search.cuh

```diff
@@ -0,0 +1,1203 @@
+      score = -score;
+    } break;
+    default: break;
+  }
+  topkScores[i + (topk * batch_ix)] = score;
+}
+
+/**
+ * @brief Compute the score for the encoded `pq_dataset` using a codebook `lut_scores`
+ *
+ * @tparam OpT an unsigned integer type that is used for bit operations on multiple PQ codes
+ *   at once; it's selected to maximize throughput while matching criteria:
+ *     1. `pq_bits * vec_len % 8 * sizeof(OpT) == 0`.
+ *     2. `pq_dim % vec_len == 0`
+ *
+ * @tparam IdxT indexing type
+ * @tparam LutT type of the elements in the lookup table.
+ *
+ * @param pq_bits The bit length of an encoded vector element after compression by PQ
+ * @param vec_len == 8 * sizeof(OpT) / gcd(8 * sizeof(OpT), pq_bits)
+ * @param pq_dim
+ * @param data_ix
+ * @param[in] pq_dataset a device pointer to the dataset [n_rows, pq_dim * pq_bits / 8]
+ * @param[in] lut_scores a device or shared memory pointer to the lookup table [pq_dim, pq_width]
+ *
+ * @return the score for the entry `data_ix` in the `pq_dataset`.
+ */
+template <typename OpT, typename IdxT, typename LutT>
+__device__ auto ivfpq_compute_score(uint32_t pq_bits,
+                                    uint32_t vec_len,
+                                    uint32_t pq_dim,
+                                    IdxT data_ix,
+                                    const uint8_t* pq_dataset,
+                                    const LutT* lut_scores) -> float
+{
+  float score                   = 0.0;
+  constexpr uint32_t kBitsTotal = 8 * sizeof(OpT);
+  const OpT* pq_head =
+    reinterpret_cast<const OpT*>(pq_dataset + uint64_t(data_ix) * (pq_dim * pq_bits / 8));
+  for (int j = 0; j < pq_dim; j += vec_len) {
+    OpT pq_code = pq_head[0];
+    pq_head += 1;
+    auto bits_left = kBitsTotal;
+    for (int k = 0; k < vec_len; k += 1) {
+      uint8_t code = pq_code;
+      if (bits_left > pq_bits) {
+        pq_code >>= pq_bits;
+        bits_left -= pq_bits;
+      } else {
+        if (k < vec_len - 1) {
+          pq_code = pq_head[0];
+          pq_head += 1;
+        }
+        code |= (pq_code << bits_left);
+        pq_code >>= (pq_bits - bits_left);
+        bits_left += (kBitsTotal - pq_bits);
+      }
+      code &= (1 << pq_bits) - 1;
+      score += float(lut_scores[code]);
+      lut_scores += (1 << pq_bits);
+    }
+  }
+  return score;
+}
+
+template <typename T>
+struct dummy_block_sort_t {
+  using queue_t = topk::warp_sort_immediate<WarpSize, true, T, uint32_t>;
+  __device__ dummy_block_sort_t(int k, uint8_t* smem_buf){};
+};
+
+template <int Capacity, typename T>
+struct pq_block_sort : topk::block_sort<topk::warp_sort_immediate, Capacity, true, T, uint32_t> {
+  using type = topk::block_sort<topk::warp_sort_immediate, Capacity, true, T, uint32_t>;
+};
+
+template <typename T>
+struct pq_block_sort<0, T> : dummy_block_sort_t<T> {
+  using type = dummy_block_sort_t<T>;
+};
+
+template <int Capacity, typename T>
+using block_sort_t = typename pq_block_sort<Capacity, T>::type;
```

Comment:
No, it's to make the whole block_sort structure compile into nothing with zero capacity (and also to access kDummy). That's because in the main kernel we actually have no compile time guards for whether to use block_sort struct.